nixos/nix-daemon: make checkConfig fully disable nix.conf validation

Atemu opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

I've got a special setup where the check causes issues and need an option to disable it entirely. I don't see much point in checking just for unknown settings and you wouldn't expect one to happen with `checkConfig = false`. 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

polykernel commented:
> I don't see much point in checking just for unknown settings

https://github.com/NixOS/nixpkgs/blob/9363ec256daa0e2b1e51ff124c27e96156d9837c/nixos/modules/services/misc/nix-daemon.nix#L74-L75
It is unforeseeable how errors will be categorized in new Nix versions so the error and warning categories are combined in the validation. The check for "error: unknown settings" is done because the Nix binary does not unambiguously distinguish between syntax errors and invalid data. It was chosen to match the module system behavior when undeclared options are used in the config.

In addition, the "unknown settings" warning is deterministic which is important for debugging as the Nix module is invariably blocking.

I do acknowledge that it is sometimes necessary to disable the validation phase completely and the Nix module currently does not provide this compatibility.

> you wouldn't expect one to happen with `checkConfig = false`.

You are right, I overlooked this on my part. The name of the option is misleading and does not convey what it actually does. I think it can be rewritten as ternary state enum(full, minimal, none) or something similar to reflect its real scope. What do you think?

Atemu committed:
nixos/nix-daemon: make checkConfig fully disable nix.conf validation

A new option checkAllErrors is introduced which implements the old checkConfig
toggle behaviour

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Atemu commented:
Sorry for getting back so late. What do you think of this @polykernel?

polykernel reviewed (approved):
I still think having `checkConfig` as an enum to specify the filtering level is more flexible, but I doubt we will reach sufficient complexity in validation to require it(one can argue validation should be done in another tool when this occurs). Given there are only three states, your approach of using another variable is simpler.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/621

SuperSandro2000 commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -383,8 +383,15 @@ in
+        '';
+      };
+
+      checkAllErrors = mkOption {
+        type = types.bool;
+        default = true;
+        description = lib.mdDoc ''
+          If enabled (the default), checks the nix.conf parsing for any kind of error. When disabled, only unknown settings are checked for.
         '';
       };
```
```suggestion
      checkAllErrors = mkOption {
        type = types.bool;
        default = true;
        description = lib.mdDoc ''
          If enabled (the default), checks the nix.conf parsing for any kind of error. When disabled, checks only for unknown settings.
        '';
      };
```
Maybe add a scenario when this is helpful to make it easier for people to know when to use this?

Atemu replied:
I don't actually know any scenario where you might want this but that's how the option worked before, so I retained that functionality.

SuperSandro2000 replied:
If we don't know any usecase for this setting then why do we have it at all? Just because it was always like does not mean we can't change it.

Atemu replied:
Sure, if you're motivated to find out, go ahead.

I'm not, so I opted to just retain the old behaviour as an option. It's also not like this option really costs us anything. It's not very complex and might help someone's odd use-case.

polykernel replied:
The purpose of having a check at the derivation level for unknown settings is to emulate checks for attribute names at the module system level. For instance, if every nix.conf setting is transformed to a corresponding typed module option, then

```nix
...
nix.settings = {
  some-unknown-setting = true;
};
...
```

would produce an error on evaluation because `settings.some-unknown-setting` is not defined. However, the current implementation would happily generate the nix.conf with the setting.

The benefit of freeform modules versus mapping all settings to module options is that it is easier to maintain but it comes at a cost of having less stringent type checking at evaluation time.

The reason the intermediate state of checking for unknown settings is needed is because `nix show-config` isn't hermetic. One case I recall where it is necessary to relax the strictness of the check is if the system is not connected to the network, `nix show-config` would output "warning: you don't have Internet access; disabling some network-dependent features" which gets converted to an error, blocking the derivation build.

pennae reviewed (commented):
looks reasonable, but could do with a small entry in the release notes. (`checkConfig = false` does behave noticably differently now after all)

Atemu committed:
rl-2211: document nix.checkConfig option changes

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -383,8 +383,15 @@ in
         description = lib.mdDoc ''
-          If enabled (the default), checks for data type mismatches and that Nix
-          can parse the generated nix.conf.
+          If enabled (the default), checks that Nix can parse the generated nix.conf.
```
```suggestion
          If enabled, checks that Nix can parse the generated nix.conf.
```

That is already in the search ui.

SuperSandro2000 commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -383,8 +383,15 @@ in
+        type = types.bool;
+        default = true;
+        description = lib.mdDoc ''
+          If enabled (the default), checks the nix.conf parsing for any kind of error. When disabled, checks only for unknown settings.
```
```suggestion
          If enabled, checks the nix.conf parsing for any kind of error. When disabled, checks only for unknown settings.
```

Atemu commented:
Thanks!
